packages/front-end/components/Settings/EditDataSource/DataSourcePipeline/DataSourcePipeline.tsx

```diff
@@ -0,0 +1,102 @@
+import { CSSProperties, FC, ReactNode, useCallback, useState } from "react";
+import { FaPencilAlt } from "react-icons/fa";
+import usePermissions from "@/hooks/usePermissions";
+import { checkDatasourceProjectPermissions } from "@/services/datasources";
+import { DataSourceQueryEditingModalBaseProps } from "../types";
+import { EditDataSourcePipeline } from "./EditDataSourcePipeline";
+
+type DataSourcePipelineProps = DataSourceQueryEditingModalBaseProps;
+
+const DataSourcePipelineField: FC<{
+  title?: string;
+  style?: CSSProperties;
+  children?: ReactNode;
+  titleClassName?: string;
+}> = ({ children, title = "", titleClassName = "", style }) => {
+  return (
+    <div className={`mb-2 ma-5 ${titleClassName}`} style={style}>
+      {title}
+      {children}
+    </div>
+  );
+};
+
+export default function DataSourcePipeline({
+  dataSource,
+  onSave,
+  canEdit,
+}: DataSourcePipelineProps) {
+  const [uiMode, setUiMode] = useState<"view" | "edit">("view");
+
+  const handleCancel = useCallback(() => {
+    setUiMode("view");
+  }, []);
+  const permissions = usePermissions();
+  canEdit =
+    canEdit &&
+    checkDatasourceProjectPermissions(
+      dataSource,
+      permissions,
+      "editDatasourceSettings"
+    );
+
+  return (
+    <div>
+      <div className="d-flex justify-content-between align-items-center mb-2">
+        <div className="d-flex justify-space-between align-items-center">
+          <h3>Data Pipeline Settings</h3>
+          <span className="badge badge-purple text-uppercase mx-2">Beta</span>
+        </div>
+        {canEdit && (
+          <div className="">
+            <button
+              className="btn btn-outline-primary font-weight-bold"
+              onClick={() => {
+                setUiMode("edit");
+              }}
+            >
+              <FaPencilAlt className="mr-1" /> Edit
+            </button>
+          </div>
+        )}
+      </div>
+      <div className="alert alert-info">
+        In this section, you can configure how GrowthBook can use write
+        permissions to your Data Source to improve the performance of experiment
+        queries.
+      </div>
+      <div>
+        <DataSourcePipelineField
+          title="Pipeline Mode: "
+          titleClassName="font-weight-bold"
+        >
+          {dataSource.settings.pipelineSettings?.allowWriting
```

Comment:
You use `dataSource.settings.pipelineSettings` a bunch in this component.  Might be worth pulling this into a separate variable at the top.  Will make the rest easier to read